version skipping WIP

RaduBerinde opened the pull request:
#### clusterversion: cockroach versions cleanup

This commit cleans up the code and constants in
`cockroach_versions.go`:
 - we separate out the "dev offset" code and improve the
   documentation;
 - we replace the key-to-version slice with a table (array). This
   makes the declaration much cleaner and makes the lookup easier.
 - we replace `ByKey()` with a method, to make the call site less
   messy; e.g. `clusterversion.ByKey(clusterversion.V23_2)` becomes
   `clusterversion.V23_2.Version()`;
 - we deprecate `BinaryMinSupportedVersionKey` in favor of `MinSupported`;
 - we deprecate `BinaryVersionKey` in favor of `Latest`;
 - we remove `roachpb.Version` "constants" in favor of using the
   `Version()` method.

Informs: #112629
Release note: None

#### WIP

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/113695)
<!-- Reviewable:end -->

RaduBerinde committed:
cli: conditionally allow version skipping

Currently the code allows upgrading from TWO previous releases (23.1
and 23.2); however, for now we only want to allow skipping a version
experimentally.

This commit changes the cli code to set the minimum supported version
to `PreviousRelease` by default, or to `MinSupported` when a special
env var is used.

Epic: none
Release note: None
